catalog/app/components/Preview/loaders/Json.js

```diff
@@ -15,22 +17,58 @@ const SCHEMA_RE = /"\$schema":\s*"https:\/\/vega\.github\.io\/schema\/([\w-]+)\/
 
 const map = (fn) => R.ifElse(Array.isArray, R.map(fn), fn)
 
+const traverseUrls = (fn, spec) => R.evolve({ data: map(R.evolve({ url: fn })) }, spec)
+
 function useVegaSpecSigner(handle) {
-  const sign = AWS.Signer.useResourceSigner()
-  // eslint-disable-next-line react-hooks/exhaustive-deps
-  return React.useCallback(
-    R.evolve({
-      data: map(
-        R.evolve({
-          url: (url) =>
-            sign({
-              ptr: Resource.parse(url),
-              ctx: { type: Resource.ContextType.Vega(), handle },
-            }),
+  const sign = AWS.Signer.useS3Signer()
+  const resolveLogicalKey = useLogicalKeyResolver()
+
+  const resolvePath = React.useMemo(
+    () =>
+      resolveLogicalKey && handle.logicalKey
+        ? (path) =>
+            resolveLogicalKey(s3paths.resolveKey(handle.logicalKey, path)).catch((e) => {
+              // eslint-disable-next-line no-console
+              console.warn(
+                `Error resolving data url '${path}' referenced from vega spec at '${handle.logicalKey}'`,
+              )
+              // eslint-disable-next-line no-console
+              console.error(e)
+              throw PreviewError.SrcDoesNotExist({ path })
+            })
+        : (path) => ({
+            bucket: handle.bucket,
+            key: s3paths.resolveKey(handle.key, path),
+          }),
+    [resolveLogicalKey, handle.logicalKey, handle.key, handle.bucket],
+  )
+
+  const processUrl = React.useMemo(
```

Comment:
well, it's more or less the same as calling function "handleClick" or smth, it tells about the type and context of the function, but to see how exactly it's being processed you have to look at the code, otherwise we should call it "maybeSignAndResolveUrl" or smth like that, which i believe is nonsense (and also we would need to adjust the name every time the logic is changed). the purpose of this function is to process the url, and the name reflects that.

Reply:
Writing code in the event handler function body is anti-pattern too, imho :)
I can let it slide because of laziness or lack of time or lack of a good idea for naming, but I can't defend it as the right approach

Instead of
```js
const handleClick = () => {
 if (a) {
   c = 2 + 2
 }
 if (b) {
  window.fetch('/api/result')
 }
}
```
one should write
```js
const handleClick = () => {
  if (a) {
    doCalculations()
  }
  if (b) {
    fetchResult()
  }
}
```